doc/man/man1/getinfo.1
.\" RCSid "$Id: getinfo.1,v 1.11 2022/03/21 17:12:49 greg Exp $"
.TH GETINFO 1 1/15/99 RADIANCE
.SH NAME
getinfo - get header information from a RADIANCE file
.SH SYNOPSIS
.B getinfo
[
.B -d
|
.B +d
][
.B "file .."
]
.br
.B getinfo
[
.B -d
]
-
.br
.B getinfo
{
.B -a
|
.B -r
}
.B "info_line1 .."
.br
.B getinfo
[
.B -d
|
.B +d
]
.B -c
.B "command .."
.SH DESCRIPTION
.I Getinfo
reads the header of each RADIANCE
.I file
and writes it to the standard output.
Octree and picture files are in a binary format, which makes
it difficult to determine their content.
Therefore, a few lines of text are placed at the beginning
of each file by the RADIANCE program that creates it.
The end of the header information and the start of the
data is indicated by an empty line.
The
.I \-d
option can be used to print the dimensions of an octree or
picture file instead.
For an octree,
.I "getinfo \-d"
prints the bounding cube (xmin ymin zmin size).
For a picture,
.I "getinfo \-d"
prints the y and x resolution (\-Y yres +X xres).
The
.I \+d
option prints the header followed by a newline and the dimensions.
If no
.I file
is given, the standard input is read.
.PP
The second form of
.I getinfo
with a hyphen simply removes the header
and copies the body of the file from the standard
input to the standard output.
If the
.I \-d
option is given, then the resolution string will be elided as well.
.PP
The third form of
.I "getinfo \-a"
adds one or more header lines to the standard input.
These lines are given as arguments to
.I getinfo,
and must be quoted if they contain spaces.
The alternative form with
.I \-r
replaces any previous occurances of the settings in the given
line(s) with nothing if there is only one word, or with
the whole line if quoted and containing interior white space or
an equals sign ('=').
Any leading white space matches any leading number of spaces and
tabs in the input header.
.PP
The fourth form of
.I "getinfo \-c"
is followed by a command and its arguments, which is executed
on the data segment of the standard input.
The header is passed along, with the addition of the command
at the end.
This is roughly equivalent to the following sequence, but does not
require the input to be in a file:
.IP "" .2i
( getinfo < input ; getinfo - < input | command .. )
.PP
If the
.I \+d
option is given with
.I \-c,
then the resolution string will be copied before the command is called.
If the
.I \-d
option is given, then the resolution string is read in but not copied.
.SH EXAMPLES
To print the header information from scene1.oct and scene2.hdr:
.IP "" .2i
getinfo scene1.oct scene2.hdr
.PP
To print the dimensions of a set of pictures:
.IP "" .2i
getinfo -d *.hdr
.PP
To add two header lines modifying the view
and exposure calibration of a picture:
.IP "" .2i
getinfo -a "VIEW= -vta -vh 180 -vv 180" "EXPOSURE=1.5" < orig.hdr > modified.hdr
.PP
To remove any indented lines containing "EXPOSURE=" and replace any
left-justified "EXPOSURE=" lines with a single value of 16:
.IP "" .2i
getinfo -r " EXPOSURE=" "EXPOSURE=16" < orig.hdr > fixed.hdr
.PP
To raise floating-point matrix elements to a power of 2.7:
.IP "" .2i
getinfo -c rcalc -if -of -e '$1=$1^2.7' < input.fmx > output.fmx
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
oconv(1), pfilt(1), rcalc(1), rhinfo(1), rpict(1), rsplit(1), rtrace(1), rvu(1)


src/util/getinfo.c
#ifndef lint
static const char	RCSid[] = "$Id: getinfo.c,v 2.24 2022/03/21 17:11:29 greg Exp $";
#endif
/*
 *  getinfo.c - program to read info. header from file.
 *
 *     1/3/86
 */

#include  <ctype.h>
#include  "rtio.h"
#include  "platform.h"
#include  "rtprocess.h"
#include  "resolu.h"

static char	fmt[MAXFMTLEN] = "*";


static int
tabstr(				/* put out line followed by tab */
	char  *s,
	void *p
)
{
	while (*s) {
		putchar(*s);
		s++;
	}
	if (*--s == '\n')
		fputc('\t', stdout);
	return(0);
}


static int
adjheadline(			/* check for lines to remove */
	char  *s,
	void *p
)
{
	char	**av;

	if (formatval(fmt, s))
		return(0);	/* don't echo format */

				/* check for match to skip */
	for (av = (char **)p; *av; av++) {
		char	*s1 = s;
		char	*s2 = *av;
		if (isspace(*s2)) {
			if (!isspace(*s1))
				continue;
			while (isspace(*++s1)) ;
			while (isspace(*++s2)) ;
		}
		while (*s2 && *s1 == *s2) {
			if ((*s1 == '=') & (s1 > s))
				return(0);
			if (isspace(*s2))
				break;
			s1++; s2++;
		}
		if (s1 == s)
			continue;
		if (isspace(*s1) && !*s2 | isspace(*s2))
			return(0);	/* skip */
	}
	fputs(s, stdout);	/* copy if no match */
	return(0);
}


static void
getdim(				/* get dimensions from file */
	FILE  *fp
)
{
	int  j;
	int  c;

	switch (c = getc(fp)) {
	case '+':		/* picture */
	case '-':
		do
			putchar(c);
		while (c != '\n' && (c = getc(fp)) != EOF);
		break;
	case 1:			/* octree */
		getc(fp);
		j = 0;
		while ((c = getc(fp)) != EOF)
			if (c == 0) {
				if (++j >= 4)
					break;
				fputc(' ', stdout);
			} else {
				putchar(c);
			}
		fputc('\n', stdout);
		break;
	default:		/* ??? */
		fputs("unknown file type\n", stdout);
		break;
	}
}


static void
copycat(void)			/* copy input to output */
{
	char	buf[8192];
	int	n;

	fflush(stdout);
	while ((n = fread(buf, 1, sizeof(buf), stdin)) > 0)
		if (writebuf(fileno(stdout), buf, n) != n)
			break;
}


int
main(
	int  argc,
	char  **argv
)
{
	int  dim = 0;
	FILE  *fp;
	int  i;

	if (argc > 1 && (argv[1][0] == '-') | (argv[1][0] == '+') &&
			argv[1][1] == 'd') {
		dim = 1 - 2*(argv[1][0] == '-');
		argc--; argv++;
	}
#ifdef getc_unlocked				/* avoid lock/unlock overhead */
	flockfile(stdin);
	flockfile(stdout);
#endif
	SET_FILE_BINARY(stdin);
	if (argc > 2 && !strcmp(argv[1], "-c")) {
		SET_FILE_BINARY(stdout);
		setvbuf(stdin, NULL, _IONBF, 2);
		if (checkheader(stdin, fmt, stdout) < 0) {
			fputs("Bad header!\n", stderr);
			return 1;
		}
		printargs(argc-2, argv+2, stdout);
		if (fmt[0] != '*')		/* better be the same! */
			fputformat(fmt, stdout);
		fputc('\n', stdout);
		if (dim) {			/* copy resolution string? */
			RESOLU	rs;
			if (!fgetsresolu(&rs, stdin)) {
				fputs("No resolution string!\n", stderr);
				return 1;
			}
			if (dim > 0)
				fputsresolu(&rs, stdout);
		}
		fflush(stdout);
		execvp(argv[2], argv+2);
		perror(argv[2]);
		return 1;
	}
	if (argc > 2 && !strcmp(argv[1], "-a")) {
		SET_FILE_BINARY(stdout);
		if (checkheader(stdin, fmt, stdout) < 0) {
			fputs("Bad header!\n", stderr);
			return 1;
		}
		for (i = 2; i < argc; i++) {
			int	len = strlen(argv[i]);
			if (!len) continue;
			fputs(argv[i], stdout);
			if (argv[i][len-1] != '\n')
				fputc('\n', stdout);
		}
		if (fmt[0] != '*')
			fputformat(fmt, stdout);
		fputc('\n', stdout);
		copycat();
		return 0;
	}
	if (argc > 2 && !strcmp(argv[1], "-r")) {
		SET_FILE_BINARY(stdout);
		if (getheader(stdin, adjheadline, argv+2) < 0) {
			fputs("Bad header!\n", stderr);
			return 1;
		}
		for (i = 2; i < argc; i++) {	/* add lines w/ var[= ]value */
			int	len = strlen(argv[i]);
			int	strt = 0, j;
			while(isspace(argv[i][strt])) strt++;
			if (strt == len) continue;
			while (isspace(argv[i][len-1])) len--;
			for (j = strt+1; j < len-1; j++)
				if (argv[i][j] == '=' || isspace(argv[i][j])) {
					for (j = 0; j < len; j++)
						putchar(argv[i][j]);
					fputc('\n', stdout);
					break;
				}
		}
		if (fmt[0] != '*')
			fputformat(fmt, stdout);
		fputc('\n', stdout);
		copycat();
		return 0;
	}
	if (argc == 2 && !strcmp(argv[1], "-")) {
		SET_FILE_BINARY(stdout);
		if (getheader(stdin, NULL, NULL) < 0) {
			fputs("Bad header!\n", stderr);
			return 1;
		}
		if (dim < 0) {			/* skip resolution string? */
			RESOLU	rs;
			if (!fgetsresolu(&rs, stdin)) {
				fputs("No resolution string!\n", stderr);
				return 1;
			}
		}
		copycat();
		return 0;
	}
	for (i = 1; i < argc; i++) {
		fputs(argv[i], stdout);
		if ((fp = fopen(argv[i], "r")) == NULL)
			fputs(": cannot open\n", stdout);
		else {
#ifdef getc_unlocked				/* avoid lock/unlock overhead */
			flockfile(fp);
#endif
			if (dim < 0) {			/* dimensions only */
				if (getheader(fp, NULL, NULL) < 0) {
					fputs("bad header!\n", stdout);
					continue;	
				}
				fputs(": ", stdout);
				getdim(fp);
			} else {
				tabstr(":\n", NULL);
				if (getheader(fp, tabstr, NULL) < 0) {
					fputs(argv[i], stderr);
					fputs(": bad header!\n", stderr);
					return 1;
				}
				fputc('\n', stdout);
				if (dim > 0) {
					fputc('\t', stdout);
					getdim(fp);
				}
			}
			fclose(fp);
		}
	}
	if (argc == 1) {
		if (dim < 0) {
			if (getheader(stdin, NULL, NULL) < 0) {
				fputs("Bad header!\n", stderr);
				return 1;	
			}
			getdim(stdin);
		} else {
			if (getheader(stdin, (gethfunc *)fputs, stdout) < 0) {
				fputs("Bad header!\n", stderr);
				return 1;
			}
			fputc('\n', stdout);
			if (dim > 0)
				getdim(stdin);
		}
	}
	return 0;
}
